### PR TITLE
Curved mesh fd

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

Only new commits are:
- Add span overload to IrregularInterpolant
- Use interpolators in SliceData

Depends on:
- [x] #5456 

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/NumericalAlgorithms/Interpolation/IrregularInterpolant.hpp [hunk 1]
```diff
@@ -68,6 +68,10 @@ class Irregular {
   DataVector interpolate(const DataVector& input) const;
   /// @}
 
+  /// \brief Interpolate a multiple variables on the grid to the target points.
```
Comment (kidder) [inline src/NumericalAlgorithms/Interpolation/IrregularInterpolant.hpp]: remove " a"

**Comment by nilsdeppe** [comment]
rebased onto the other PR.

File: src/Evolution/DgSubcell/SliceTensor.hpp [hunk 2]
```diff
@@ -39,24 +45,30 @@ void slice_tensor_for_subcell(
  * \brief Slice a single volume tensor for a given direction and slicing depth
  * (number of ghost points).
  *
- * Note that the last argument has the type `Direction<Dim>`, not a DirectionMap
- * (cf. `evolution::dg::subcell::slice_data`)
+ * Note that the second last argument has the type `Direction<Dim>`, not a
```
Comment (kidder) [inline src/Evolution/DgSubcell/SliceTensor.hpp]: second to

File: tests/Unit/Evolution/DgSubcell/Test_PrepareNeighborData.cpp [hunk 3]
```diff
@@ -324,9 +359,12 @@ void test(const bool all_neighbors_are_doing_dg,
...
                   "[Evolution][Unit]") {
   for (const auto& [all_neighbors_are_doing_dg, fd_deriv_order] :
-       cartesian_product(make_array(true, false),
-                         make_array(::fd::DerivativeOrder::Two,
-                                    ::fd::DerivativeOrder::Four))) {
+       cartesian_product(
+           std::array{true, false},
+           std::array{
+               ::fd::DerivativeOrder::Two  // ,
+                                           // ::fd::DerivativeOrder::Four
```
Comment (kidder) [inline tests/Unit/Evolution/DgSubcell/Test_PrepareNeighborData.cpp]: remove

File: tests/Unit/Evolution/DgSubcell/Test_SliceTensor.cpp [hunk 4]
```diff
@@ -55,14 +56,37 @@ void test() {
...
+              volume_extents.slice_away(direction.dimension()).product() *
+                  num_ghost_pts,
+              10.0};
+          Scalar<DataVector> sliced_scalar = expected_sliced_scalar;
+          evolution::dg::subcell::slice_tensor_for_subcell(
+              make_not_null(&sliced_scalar), volume_scalar, volume_extents,
+              num_ghost_pts, direction, fd_to_neighbor_fd_interpolants);
+#ifdef SPECTRE_DEBUG
+          disable_floating_point_exceptions();
+          CHECK(sliced_scalar != expected_sliced_scalar);
```
Comment (kidder) [inline tests/Unit/Evolution/DgSubcell/Test_SliceTensor.cpp]: I don't understand what this checks....

File: tests/Unit/Evolution/DgSubcell/Test_SliceVariable.cpp [hunk 5]
```diff
@@ -34,9 +34,34 @@ void test_slice(
...
+                num_ghost_pts,
+            10};
+    Variables<tmpl::list<Tags::Scalar, Tags::TensorI<Dim>>> sliced_vars{
+        expected_sliced_vars};
+    evolution::dg::subcell::slice_variable(
+        make_not_null(&sliced_vars), volume_vars, volume_extents, num_ghost_pts,
+        direction, fd_to_neighbor_fd_interpolants);
+#ifdef SPECTRE_DEBUG
+    disable_floating_point_exceptions();
+    CHECK(sliced_vars != expected_sliced_vars);
```
Comment (kidder) [inline tests/Unit/Evolution/DgSubcell/Test_SliceVariable.cpp:62]: I don't understand what this is checking

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
Done, rebased, squashed. Thanks for the review! Here's the patch for the second commit.

```patch
diff --git a/src/Evolution/DgSubcell/SliceTensor.hpp b/src/Evolution/DgSubcell/SliceTensor.hpp
index 0809d9f025..5f72bb42b7 100644
--- a/src/Evolution/DgSubcell/SliceTensor.hpp
+++ b/src/Evolution/DgSubcell/SliceTensor.hpp
@@ -45,7 +45,7 @@ void slice_tensor_for_subcell(
  * \brief Slice a single volume tensor for a given direction and slicing depth
  * (number of ghost points).
  *
- * Note that the second last argument has the type `Direction<Dim>`, not a
+ * Note that the second to last argument has the type `Direction<Dim>`, not a
  * DirectionMap (cf. `evolution::dg::subcell::slice_data`)
  *
  */
diff --git a/tests/Unit/Evolution/DgSubcell/Test_PrepareNeighborData.cpp b/tests/Unit/Evolution/DgSubcell/Test_PrepareNeighborData.cpp
index d119cf3d23..a32f1f0ae8 100644
--- a/tests/Unit/Evolution/DgSubcell/Test_PrepareNeighborData.cpp
+++ b/tests/Unit/Evolution/DgSubcell/Test_PrepareNeighborData.cpp
@@ -359,12 +359,9 @@ void test(const bool all_neighbors_are_doing_dg,
 SPECTRE_TEST_CASE("Unit.Evolution.Subcell.PrepareNeighborData",
                   "[Evolution][Unit]") {
   for (const auto& [all_neighbors_are_doing_dg, fd_deriv_order] :
-       cartesian_product(
-           std::array{true, false},
-           std::array{
-               ::fd::DerivativeOrder::Two  // ,
-                                           // ::fd::DerivativeOrder::Four
-           })) {
+       cartesian_product(std::array{true, false},
+                         std::array{::fd::DerivativeOrder::Two,
+                                    ::fd::DerivativeOrder::Four})) {
     test<1>(all_neighbors_are_doing_dg, fd_deriv_order);
     test<2>(all_neighbors_are_doing_dg, fd_deriv_order);
     test<3>(all_neighbors_are_doing_dg, fd_deriv_order);
diff --git a/tests/Unit/Evolution/DgSubcell/Test_SliceTensor.cpp b/tests/Unit/Evolution/DgSubcell/Test_SliceTensor.cpp
index 04dd28b653..e4943ee8f7 100644
--- a/tests/Unit/Evolution/DgSubcell/Test_SliceTensor.cpp
+++ b/tests/Unit/Evolution/DgSubcell/Test_SliceTensor.cpp
@@ -74,7 +74,11 @@ void test() {
               num_ghost_pts, direction, fd_to_neighbor_fd_interpolants);
 #ifdef SPECTRE_DEBUG
           disable_floating_point_exceptions();
-          CHECK(sliced_scalar != expected_sliced_scalar);
+          CHECK(get(sliced_scalar).size() ==
+                get(expected_sliced_scalar).size());
+          for (const double value_at_point : get(sliced_scalar)) {
+            CHECK(isnan(value_at_point));
+          }
           enable_floating_point_exceptions();
 #endif
         }
diff --git a/tests/Unit/Evolution/DgSubcell/Test_SliceVariable.cpp b/tests/Unit/Evolution/DgSubcell/Test_SliceVariable.cpp
index bc171d7d71..b1d2f53dbc 100644
--- a/tests/Unit/Evolution/DgSubcell/Test_SliceVariable.cpp
+++ b/tests/Unit/Evolution/DgSubcell/Test_SliceVariable.cpp
@@ -54,6 +54,11 @@ void test_slice(
         direction, fd_to_neighbor_fd_interpolants);
 #ifdef SPECTRE_DEBUG
     disable_floating_point_exceptions();
+    CHECK(sliced_vars.size() == expected_sliced_vars.size());
+    for (size_t i = 0; i < sliced_vars.size(); ++i) {
+      CHECK(isnan(
+          *std::next(sliced_vars.data(), static_cast<std::ptrdiff_t>(i))));
+    }
     CHECK(sliced_vars != expected_sliced_vars);
     enable_floating_point_exceptions();
 #endif

```

**Review by kidder — DISMISSED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**